GradienceTeam/Gradience
# GNOME Shell theming support

# Description

#### THE SHELL THEMING IS REAL!!!
This is it, Shell theming finally comes to Gradience. This means that we only need to add support to additional _420_ frameworks and terminal emulators for a "complete GNOME themeing beauty" (sic).

But in all seriousness, this PR brings up shell theming support for GNOME Shell versions 43 and 42. It is WIP, but can be tested already with a majority of available presets (~excluding light presets~ Light presets support was significantly improved in commit 7e9ac2c4535dc3fcc17e5b1c994ccdf545294bcd and can be tested as well).

This PR also introduces a new concept of Theme Engines and improves designs of Theming tab and Preferences window, to make them more readable and expandable.

### To-do list:

- [x] Models for storing shell theme properties
- [x] Template system for variables insertion
- [x] Bare-minimum functions for variables insertion, SCSS compiling and theme apply
- [x] Automatic theme selection (no need to manually recheck Gradience theme in extension)
- [x] ~Make SCSS compilation multi-threaded~ (We don't need that, as `libsass` is pretty quick at compiling)
- [x] Recolor Shell panel
- [x] Recolor Shell palette
- [x] Convert palette values from GTK to hex/rgba color formats
- [x] Add support for custom CSS in Advanced tab
- [x] Check if host has GNOME installed (Needs testing)
- [x] Make current interface less confusing
- [x] Copy theme source files to user-writable directory
- [x] UI for theme manipulation and apply
   - [x] UI for applying shell themes
   - [x] UI for setting custom colors
   - [x] Shell theme apply mechanism in backend code
   - [x] Custom colors insertion mechanism in backend code
- [x] Shell theme source code in submodules
- [x] Colored switches (see: https://github.com/avanishsubbiah/material-you-theme/issues/41)
- [x] Shell theming via command line
- [x] ~System-wide theming~ (Reconsidered for the future releases)
- [x] Flatpak support

### Screenshots
![screen-shell-1](https://user-images.githubusercontent.com/73042332/209346301-91b7b08e-06ef-4af0-8205-605464aae7f8.png)
<figcaption>Full desktop view with a theme that goes hard</figcaption>
<hr>

![screen-shell-3](https://user-images.githubusercontent.com/73042332/209346985-d70a0c35-a9b7-4486-b2ec-48f6e46acc94.png)
<figcaption>OSD widgets properly themed!? No way!</figcaption>
<hr>

![screen-shell-2](https://user-images.githubusercontent.com/73042332/209347127-01a7c2ae-027d-44c9-b3ac-8cb6aa8d7dbc.png)
<figcaption>Nice</figcaption>
<hr>




Fixes #618 and https://github.com/GradienceTeam/Plugins/issues/11

## Type of change



> **Warning** This PR will require a change in documentation

## Changelog 

- Changed **Monet** section name to **Theming**
- `GradienceOptionRow` widget is now more flexible
- Make **Theming** section more modular
- Added new group in **Theming** section
- Renamed `settings_schema` to `preset_schema` and moved to `frontend/schemas` subdir
- Added new _GNOME Shell_ option in `GradienceAppTypeDialog` view
- Created new model module `ShellTheme` with available manipulation functions

## Testing


### How to test the changes

Clone this repo with `--recurse-submodules` flag: `git clone --recurse-submodules https://github.com/GradienceTeam/Gradience.git` and switch branches to get into this PR: `git checkout shell-theming`

Make sure you have `User Themes` GNOME extension installed (and enabled!), which you should as it is an built-in plugin, but check it nonetheless.

Load a preset you want to generate a theme from. Click on `Theming` tab in main window, choose in `Preset Variant` row which color variant is your preset and click `Apply` button in Shell Engine.

~Open `User Themes` extension settings and check (or recheck) `gradience-shell` theme.~ Not needed as of commit 6493453845794399800b1a276ab33b29393052e5
